Compliance with callback standards

hkan opened the pull request:
Current library binds socket server errors to the second argument of ack callback. This is not desirable as Node.js standards always use first argument as the error parameter. This PR fixes the problem.

Current code:

    socket.emit('foo', timeout((timeoutError, socketError, data) => {
      if (timeoutError || socketError) {
        // handle errors
      }

      bar(data)
    })

How it will be with this PR:

    socket.emit('foo', timeout((err, data) => {
      if (err) { // err might be a timeout error or a server failure response
        // handle errors
      }

      bar(data)
    })

hkan committed:
updated for compliance with callback standards

jakubknejzlik commented:
@hkan hi and thanks for your PR. Actually it reverts the updates in commit related to issue #3 . Socket.io seems to not follow the callback error convention (according to their docs https://socket.io/docs/server-api/#). Any thoughts about this @DmitryMyadzelets?

hkan commented:
I see. Although I never came across a socket server that employs socket.io and does not comply with Node style callbacks, I also don't think we should assume the compliance on the server-side.

Would it make sense to provide this PR's functionality under a condition - e.g. a third boolean parameter? The default behavior will not be changed for backwards compat and the third parameter (as `true`?) would skip the `args.unshift(null);` line.

jakubknejzlik commented:
That's great idea and I think it's the right solution to this problem  :-). Would you mind updating the PR? 
I'm also thinking about creating it as `options` parameter even if there's just single option right now. What do you think?

hkan commented:
Yeah, an object would be better for probable future improvements than a boolean. I will update the PR ASAP.

hkan committed:
implemented optional nodejs style callback compliance

hkan committed:
updated tests accordingly

hkan committed:
updated syntax with ES6 features

hkan commented:
Hey, @jakubknejzlik! I added the option object, along with its tests.

I also updated some lines with new ES6 features such as object spread syntax. There was a `const` keyword, and I figured anyone using this library already has to have a ES6-supporting Node engine... So, I thought why wouldn't we take full advantage of new syntax? 🤓
